_posts/2022-10-06-audio-classification-torch/audio-classification.Rmd

```diff
@@ -0,0 +1,472 @@
+
+    [1]  "bed"    "bird"   "cat"    "dog"    "down"   "eight"
+    [7]  "five"   "four"   "go"     "happy"  "house"  "left"
+    [32] " marvin" "nine"   "no"     "off"    "on"     "one"
+    [19] "right"  "seven" "sheila" "six"    "stop"   "three"
+    [25]  "tree"   "two"    "up"     "wow"    "yes"    "zero" 
+
+Picking a sample at random, we see that the information we'll need is contained in four properties: `waveform`, `sample_rate`, `label_index`, and `label`.
+
+The first, `waveform`, will be our predictor.
+
+```{r}
+sample <- ds[2000]
+dim(sample$waveform)
+```
+
+    [1]     1 16000
+
+Individual tensor values are centered at zero, and range between -1 and 1. There are 16,000 of them, reflecting the fact that the recording lasted for one second, and was registered at (or has been converted to, by the dataset creators) a rate of 16,000 samples per second. The latter information is stored in `sample$sample_rate`:
+
+```{r}
+sample$sample_rate
+```
+
+    [1] 16000
+
+All recordings have been sampled at the same rate. Their length almost always equals one second; the -- very -- few sounds that are minimally longer we can safely truncate.
+
+Finally, the target is stored, in integer form, in `sample$label_index`, the corresponding word being available from `sample$label`:
+
+```{r}
+sample$label
+sample$label_index
+```
+
+    [1] "bird"
+    torch_tensor
+    2
+    [ CPULongType{} ]
+
+How does this audio signal "look"?
+
+```{r}
+library(ggplot2)
+
+df <- data.frame(
+  x = 1:length(sample$waveform[1]),
+  y = as.numeric(sample$waveform[1])
+  )
+
+ggplot(df, aes(x = x, y = y)) +
+  geom_line(size = 0.3) +
+  ggtitle(
+    paste0(
+      "The spoken word \"", sample$label, "\": Sound wave"
+    )
+  ) +
+  xlab("time") +
+  ylab("amplitude") +
+  theme_minimal()
+```
+
+![The spoken word "bird", in time-domain representation.](images/audio-bird-waveform.png){#fig-audio-bird-waveform fig-alt="A sound wave, displaying amplitude over time."}
+
+What we see is a sequence of amplitudes, reflecting the sound wave produced by someone saying "bird". Put differently, we have here a time series of "loudness values". Even for experts, guessing *which* word resulted in those amplitudes is an impossible task. This is where domain knowledge comes in. The expert may not be able to make much of the signal *in this representation*; but they may know a way to more meaningfully represent it.
+
+## Two equivalent representations
+
+Imagine that instead of as a sequence of amplitudes over time, the above wave were represented in a way that had no information about time at all. Next, imagine we took that representation and tried to recover the original signal. For that to be possible, the new representation would somehow have to contain "just as much" information as the wave we started from. That "just as much" is obtained from the *Fourier Transform*, and it consists of the magnitudes and phase shifts of the different *frequencies* that make up the signal.
```

Comment:
> the above wave were represented

I'm not sure if you want plural ("waves were") or singular ("wave was"), but the noun and verb should agree, either way.

Otherwise, really nice explanation.

Reply:
it's supposed to be conjunctive (I like the _were_ ;-))